mlflow/utils/download_cloud_file_chunk.py

```diff
@@ -0,0 +1,195 @@
+        _pid=os.getpid(),
+    )
+
+
+def cloud_storage_http_request(
+    method,
+    url,
+    max_retries=5,
+    backoff_factor=2,
+    retry_codes=_TRANSIENT_FAILURE_RESPONSE_CODES,
+    timeout=None,
+    **kwargs,
+):
+    """
+    Performs an HTTP PUT/GET/PATCH request using Python's `requests` module with automatic retry.
+
+    :param method: string of 'PUT' or 'GET' or 'PATCH', specify to do http PUT or GET or PATCH
+    :param url: the target URL address for the HTTP request.
+    :param max_retries: maximum number of retries before throwing an exception.
+    :param backoff_factor: a time factor for exponential backoff. e.g. value 5 means the HTTP
+      request will be retried with interval 5, 10, 20... seconds. A value of 0 turns off the
+      exponential backoff.
+    :param retry_codes: a list of HTTP response error codes that qualifies for retry.
+    :param timeout: wait for timeout seconds for response from remote server for connect and
+      read request. Default to None owing to long duration operation in read / write.
+    :param kwargs: Additional keyword arguments to pass to `requests.Session.request()`
+
+    :return requests.Response object.
+    """
+    if method.lower() not in ("put", "get", "patch", "delete"):
+        raise ValueError("Illegal http method: " + method)
+    return _get_http_response_with_retries(
+        method,
+        url,
+        max_retries,
+        backoff_factor,
+        retry_codes,
+        timeout=timeout,
+        **kwargs,
+    )
+
+
+def _get_http_response_with_retries(
+    method, url, max_retries, backoff_factor, retry_codes, **kwargs
+):
+    """
+    Performs an HTTP request using Python's `requests` module with an automatic retry policy.
+
+    :param method: a string indicating the method to use, e.g. "GET", "POST", "PUT".
+    :param url: the target URL address for the HTTP request.
+    :param max_retries: Maximum total number of retries.
+    :param backoff_factor: a time factor for exponential backoff. e.g. value 5 means the HTTP
+      request will be retried with interval 5, 10, 20... seconds. A value of 0 turns off the
+      exponential backoff.
+    :param retry_codes: a list of HTTP response error codes that qualifies for retry.
+    :param kwargs: Additional keyword arguments to pass to `requests.Session.request()`
+
+    :return: requests.Response object.
+    """
+    session = _get_request_session(max_retries, backoff_factor, retry_codes)
+    return session.request(method, url, **kwargs)
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--range-start", required=True, type=int)
+    parser.add_argument("--range-end", required=True, type=int)
+    parser.add_argument("--headers", required=True, type=str)
+    parser.add_argument("--download-path", required=True, type=str)
+    parser.add_argument("--http-uri", required=True, type=str)
+    parser.add_argument("--temp-file", required=True, type=io.TextIOWrapper)
```

Comment:
Does `type=io.TextIOWrapper` really work?

```python
import argparse
import io


p = argparse.ArgumentParser()
p.add_argument("--file", type=io.TextIOWrapper)
p.parse_args()
```

gives:

```
> python a.py --file foo
Traceback (most recent call last):
  File "a.py", line 7, in <module>
    p.parse_args()
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.8/argparse.py", line 1768, in parse_args
    args, argv = self.parse_known_args(args, namespace)
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.8/argparse.py", line 1800, in parse_known_args
    namespace, args = self._parse_known_args(args, namespace)
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.8/argparse.py", line 2006, in _parse_known_args
    start_index = consume_optional(start_index)
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.8/argparse.py", line 1946, in consume_optional
    take_action(action, args, option_string)
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.8/argparse.py", line 1858, in take_action
    argument_values = self._get_values(action, argument_strings)
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.8/argparse.py", line 2389, in _get_values
    value = self._get_value(action, arg_string)
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.8/argparse.py", line 2422, in _get_value
    result = type_func(arg_string)
AttributeError: 'str' object has no attribute 'readable'
```

Reply:
What is foo in your test?

Reply:
foo is just a random string

Reply:
Does a file named foo need to exist?

Reply:
I don't think it's a string, it's of type io.TextIOWrapper. So when I use it, I pass `with tempfile.TemporaryFile(mode="w+") as temp_file` temp_file in, which is of this type.

Reply:
```python
# this should be equivalent to what `_exec_cmd` does. 

import subprocess
import tempfile

with tempfile.TemporaryFile(mode="w+") as temp_file:
    subprocess.run(["cat", str(temp_file)], check=True)
```

fails with:

```
cat: '<_io.TextIOWrapper name=3 mode='\''w+'\'' encoding='\''UTF-8'\''>': No such file or directory
Traceback (most recent call last):
  File "b.py", line 5, in <module>
    subprocess.run(["cat", str(temp_file)], check=True)
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.8/subprocess.py", line 516, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['cat', "<_io.TextIOWrapper name=3 mode='w+' encoding='UTF-8'>"]' returned non-zero exit status 1.
```

Reply:
I think you're right. So instead of writing to a file, maybe I'll just output in stdout then use regular expression to catch the specific pattern?

Reply:
We can write to a file. Regular expression is fragile and can lead to unintended results.